### PR TITLE
Convert SvgEdges to functional component

### LINKED ISSUE #3397: [UI Refactor] Migrate SvgEdges to Functional Component
This issue is part of the larger effort to modernize the Jaeger UI codebase (Parent Issue #2610). The goal is to refactor the `SvgEdges` component from a React Class Component to a React Functional Component using Hooks.

**Target Component:** `packages/plexus/src/Digraph/SvgEdges.tsx`

> **Note:** This component is part of the `plexus` package, which is a graph visualization library used by Jaeger UI.

### Acceptance Criteria

- [ ] The component is converted to a Functional Component.
- [ ] `this.state` is replaced with `useState` or `useReducer`.
- [ ] Lifecycle methods (componentDidMount, etc.) are replaced with `useEffect`.
- [ ] `memoizeOne` calls are converted to `useMemo` where appropriate.
- [ ] Static properties are handled appropriately (e.g., moved outside component or use module-level constants).
- [ ] All existing PropType validations are preserved.
- [ ] **Performance:** Wrap in `React.memo` to maintain `PureComponent` behavior.
- [ ] **Visual Verification:** Attach "Before" and "After" screenshots to the PR to prove no visual regression.
- [ ] **Tests:** Existing unit tests must pass.

### Plexus-Specific Guidance

| Pattern | Class Approach | Functional Approach |
| --- | --- | --- |
| **Instance ID** | `baseId = \`prefix--${idCounter++}\`` | `const baseId = useRef(\`prefix--${idCounter++}\`).current` |
| **Memoization** | `memoizeOne(fn)` | `useMemo(() => fn, [deps])` or keep `memoizeOne` with `useRef` |
| **Static Props** | `static defaultProps = {...}` | Function parameter defaults or separate constant |
| **Render Utils** | `this.renderUtils = {...}` | `const renderUtils = useMemo(() => ({...}), [deps])` |

### Migration Guide (Cheat Sheet)

| Feature | Old (Class) | New (Functional) |
| --- | --- | --- |
| **State** | `this.state = { isOpen: false }` | `const [isOpen, setIsOpen] = useState(false);` |
| **Side Effects** | `componentDidMount()` | `useEffect(() => { ... }, [])` |
| **Props** | `this.props.traceId` | `props.traceId` (or destructure: `const { traceId } = props;`) |
| **Default Props** | `static defaultProps = { name: 'Guest' }` | `function Component({ name = 'Guest' }) { ... }` |
| **Performance** | `class MyComp extends React.PureComponent` | `export default React.memo(MyComp)` |
| **Context** | `static contextType = MyContext` | `const val = useContext(MyContext)` |
| **Refs** | `this.myRef = React.createRef()` | `const myRef = useRef(null)` |

### Common Pitfalls (Read Carefully!)

- **Stale Closures:** Be careful when using `useEffect`. If you use a variable inside the effect that comes from props or state, it must be in the dependency array. If you omit it, the effect will use the old value.
- **Testing with Enzyme:** Functional components do not have instances. Tests like `wrapper.instance().method()` or `wrapper.state('value')` will fail. You should refactor these tests to simulate events (clicks) and check the rendered output instead.
- **Static Properties:** Class components can have static properties attached directly. For functional components, you'll need to assign these after the function definition or use module-level constants.

### Resources

- [React Official Guide: Hooks at a Glance](https://reactjs.org/docs/hooks-overview.html)
- [A Complete Guide to useEffect (Overreacted.io)](https://overreacted.io/a-complete-guide-to-useeffect/) - Essential reading for understanding dependency arrays.
- [React Redux Hooks Documentation](https://react-redux.js.org/api/hooks) - Guide for replacing `connect()`.
- [Thinking in React Hooks](https://wattenberger.com/blog/react-hooks) - Visual guide to the mental model shift.
- [How to fetch data with React Hooks](https://www.robinwieruch.de/react-hooks-fetch-data) - Handling async operations.

### PR DESCRIPTION
**Pull request by thc1006**
## Which problem is this PR solving?

Resolves #3397

## Description of the changes

Converted the `SvgEdges` class component to a functional component using React.memo with custom comparator.

## How was this change tested?

- Added unit tests for SvgEdges component
- Run `npm run test` in packages/plexus directory

## Checklist

- [x] I have read https://github.com/jaegertracing/jaeger/blob/master/CONTRIBUTING_GUIDELINES.md
- [x] I have signed all commits
- [x] I have added unit tests for the new functionality
- [x] I have run lint and test steps successfully
  - for `jaeger-ui`: `npm run lint` and `npm run test`

### CODE REVIEW

**Comment by yurishkuro** [comment]
unit tests are failing

**Comment by thc1006** [comment]
> unit tests are failing

Done!

**Comment by thc1006** [comment]
Dear @yurishkuro,

Following up on the previous fixes, I have addressed the additional feedback:

- Added complete memoization test coverage for all props in arePropsEqual (getClassName, renderUtils, markerStartId, setOnEdge)
- Restored the original Uber copyright header in SvgEdges.tsx to maintain consistency with other plexus files
- Updated test file copyright to match the plexus package convention

All 21 tests pass with 100% coverage for SvgEdges.tsx.

Regarding the latest Copilot suggestion about testing setOnEdge as an array: the current tests verify memoization behavior for function references, which covers the primary use case. The array comparison logic is an implementation detail of isSamePropSetter utility. I can add those tests if you think they are necessary for this PR.

Ready for review when you have a chance. Thank you.

File: packages/plexus/src/Digraph/SvgEdges.test.js [hunk 1]
```diff
@@ -0,0 +1,326 @@
...
+            getClassName={mockGetClassName}
+            layoutEdges={mockLayoutEdges}
+            renderUtils={mockRenderUtils}
+            setOnEdge={newSetOnEdge}
+          />
+        </svg>
+      );
+
+      expect(mockRenderCount).toBeGreaterThan(initialRenderCount);
+    });
```
[the remark this answers is not in the document]
Comment (thc1006) [reply packages/plexus/src/Digraph/SvgEdges.test.js:324]: There is no any this kinda array `isSamePropSetter` testing in whole plexus Package.
if really need may ask dear rela maintainer !

File: packages/plexus/src/Digraph/SvgEdges.tsx [hunk 2]
```diff
@@ -17,32 +17,39 @@ type TProps<T = {}> = {
...
-        layoutEdge={edge}
-        markerEndId={markerEndId}
-        markerStartId={markerStartId}
-        renderUtils={renderUtils}
-        setOnEdge={setOnEdge}
-        label={edge.edge.label}
-      />
-    ));
-  }
+// Comparison function that mirrors the original shouldComponentUpdate logic
```
Comment (yurishkuro) [inline packages/plexus/src/Digraph/SvgEdges.tsx:20]: Why can't we simply rely on React doing the comparison? That is literally the reason why React was created.

Comment (thc1006) [reply packages/plexus/src/Digraph/SvgEdges.tsx:20]: Sorry for the delayed reply, I was at KubeCon EU these past two weeks.

Thank you for the review. You're right, the custom comparator shouldn't be here. I raised the same point on #3403 and offered to simplify, but should have just gone ahead and done it. The isSamePropSetter call on setOnEdge is handling something the caller should stabilize with useMemo, not something the child should work around.

I'll switch to plain React.memo() here and on #3403. Both need a rebase against the toolchain changes anyway, so I'll handle them together.

**Comment by thc1006** [comment]
Apologies for the back and forth here. After taking a closer look at main, I realized SvgEdges.tsx was already converted to a functional component in #3518 by @hharshhsaini. I should have checked before replying to the arePropsEqual comment earlier.

Thank you for the feedback regardless, it clarified the right approach for Nodes (#3403) as well.

One thing that might still have value: this branch includes a 326-line test file for SvgEdges covering rendering, memoization behavior, and edge cases. There don't appear to be tests for this component on main yet.

Would a standalone test PR (migrated to Vitest) be welcome? Happy to put one together if so.
